std/concurrency.d

```diff
@@ -1682,13 +1683,54 @@ class Generator(T) :
 
 
     /**
-     * Returns the most recently generated value.
+     * Returns the most recently generated value by shallow copy.
      */
     final T front() @property
     {
         return *m_value;
     }
 
+    /**
+     * Returns the most recently generated value without excuting a copy
+     * contructor. Will not compile for element types defining a
+     * postblit, because Generator does not return by reference.
+     */
+    final T moveFront()
+    {
+        import std.algorithm, std.range;
+        static if (!hasElaborateCopyConstructor!T)
+        {
+            return front;
+        }
+        else
+        {
+            static assert(0,
+                    "Fiber front is rvalue and thus cannot be moved when it defines a postblit.");
+        }
+    }
+
+    final int opApply(scope int delegate(T) loopBody)
+    {
+        int broken;
+        for(; !empty; popFront())
+        {
+            broken = loopBody(front);
+            if(broken) break;
+        }
+        return broken;
+    }
+
+    final int opApply(scope int delegate(size_t, T) loopBody)
```

Comment:
> If somebody adds an opApply implementation using foreach to enumerate(), it would regress Generators opApply because it would recurse infinitely then.

I am sorry, but I can't follow. How would sth. like this half-pseudo code end in an infinite recursion?

```d
struct Enumerate(R)
{
    R r;
    int opApply(scope int delegate(size_t, ElementType!R) dg)
    {
         int err;
         auto r2 = r.save;
         for (size_t i; !r2.empty; r2.popFront(), i++)
         {
             err = dg(i, r2.front);
             if (err) return err;
         }
         return 0;
    }
}
myGenerator.enumerate.array;
```